File: apps/admin-panel/src/lib/storage.ts
import { ApiClient, apiClient } from './api';
import { buildAuthHeaders } from './http';

export interface StorageNode {
  path: string;
  type: 'folder' | 'file';
  size?: number;
  children?: StorageNode[];
}

export type TrashItemType = 'book' | 'app' | 'unknown';

export interface TrashEntry {
  key: string;
  bucket: string;
  path: string;
  item_type: TrashItemType;
  object_count: number;
  total_size: number;
  metadata?: Record<string, string>;
}

export interface RestoreResponse {
  restored_key: string;
  objects_moved: number;
  item_type: TrashItemType;
}

export interface TrashDeleteResponse {
  deleted_key: string;
  objects_removed: number;
  item_type: TrashItemType;
}

export const listAppContents = (
  platform: string,
  token: string,
  tokenType: string = 'Bearer',
  client: ApiClient = apiClient
): Promise<StorageNode> => client.get<StorageNode>(`/storage/apps/${platform}`, {
  headers: buildAuthHeaders(token, tokenType)
});

export const listTrashEntries = (
  token: string,
  tokenType: string = 'Bearer',
  client: ApiClient = apiClient
): Promise<TrashEntry[]> =>
  client.get<TrashEntry[]>('/storage/trash', {
    headers: buildAuthHeaders(token, tokenType)
  });

export const restoreTrashEntry = (
  key: string,
  token: string,
  tokenType: string = 'Bearer',
  client: ApiClient = apiClient
): Promise<RestoreResponse> =>
  client.post<RestoreResponse, { key: string }>(
    '/storage/restore',
    { key },
    { headers: buildAuthHeaders(token, tokenType) }
  );

export const deleteTrashEntry = (
  key: string,
  token: string,
  tokenType: string = 'Bearer',
  client: ApiClient = apiClient
): Promise<TrashDeleteResponse> =>
  client.delete<TrashDeleteResponse, { key: string }>(
    '/storage/trash',
    { key },
    { headers: buildAuthHeaders(token, tokenType) }
  );

File: apps/api/app/schemas/storage.py
"""Pydantic schemas for storage-related endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.book import BookRead


class TrashEntryRead(BaseModel):
    """Represents an aggregated trash entry available for restoration."""

    key: str = Field(..., description="Full key within the trash bucket including trailing slash")
    bucket: str = Field(..., description="Original bucket the content belonged to")
    path: str = Field(..., description="Original path within the bucket")
    item_type: Literal["book", "app", "unknown"] = Field(default="unknown")
    object_count: int = Field(..., ge=0)
    total_size: int = Field(..., ge=0)
    metadata: dict[str, str] | None = None
    youngest_last_modified: datetime | None = Field(
        default=None,
        description="UTC timestamp for the most recently modified object within this entry",
    )
    eligible_at: datetime | None = Field(
        default=None,
        description="UTC timestamp when the entry exits the enforced retention window",
    )
    eligible_for_deletion: bool = Field(
        default=False,
        description="Indicates whether the entry can be deleted without a retention override",
    )


class RestoreRequest(BaseModel):
    """Request payload for restoring an item from the trash bucket."""

    key: str = Field(..., description="Key identifying the trash entry to restore")


class RestoreResponse(BaseModel):
    """Response returned after initiating a restore operation."""

    restored_key: str
    objects_moved: int
    item_type: Literal["book", "app", "unknown"]
    book: BookRead | None = None


class TrashDeleteRequest(BaseModel):
    """Request payload for permanently deleting a trash entry."""

    key: str = Field(..., description="Key identifying the trash entry to delete")
    force: bool = Field(default=False, description="Bypass retention checks when true")
    override_reason: str | None = Field(
        default=None,
        description="Justification recorded when bypassing retention checks",
        max_length=500,
    )


class TrashDeleteResponse(BaseModel):
    """Response returned after a successful permanent deletion."""

    deleted_key: str
    objects_removed: int
    item_type: Literal["book", "app", "unknown"]


__all__ = [
    "TrashEntryRead",
    "RestoreRequest",
    "RestoreResponse",
    "TrashDeleteRequest",
    "TrashDeleteResponse",
]
